diff --git a/mwccgap/elf.py b/mwccgap/elf.py
index 30e56a2..e196fb0 100644
--- a/mwccgap/elf.py
+++ b/mwccgap/elf.py
@@ -197,13 +197,11 @@ class Elf:
             symbol.name = self.strtab.get_symbol_by_index(symbol.st_name)
 
         # To be refined
-        function_names = [
-            s.name
-            for s in sorted(
-                filter(lambda x: x.st_info in FUNCTION_ST_INFOS, self.symtab.symbols),
-                key=lambda x: x.st_shndx,
-            )
-        ]
+        function_names = {
+            symbol.st_shndx: symbol.name
+            for symbol in self.symtab.symbols
+            if symbol.st_info in FUNCTION_ST_INFOS and symbol.st_shndx != 0
+        }
 
         for i, section in enumerate(self.sections):
             if isinstance(section, RelocationRecord):
@@ -214,12 +212,7 @@ class Elf:
                 if section.name == ".text":
                     text_section = TextSection.from_section(section)
                     self.sections[i] = text_section
-                    if len(function_names) == 0:
-                        # FIXME: without globl label, ASM has no functions
-                        pass
-                    else:
-                        function_name = function_names[len(self.functions)]
-                        text_section.function_name = function_name
+                    text_section.function_name = function_names.get(i, "")
                     self.functions.append(text_section)
                 elif section.name == ".rodata":
                     self.rodata_sections.append(section)
